libcxx/CMakeLists.txt

```diff
@@ -293,6 +293,7 @@ option(LIBCXX_ENABLE_THREADS "Build libc++ with support for threads." ON)
 option(LIBCXX_ENABLE_MONOTONIC_CLOCK
   "Build libc++ with support for a monotonic clock.
    This option may only be set to OFF when LIBCXX_ENABLE_THREADS=OFF." ON)
+option(LIBCXX_USE_LLVM_LIBC "Build libc++ against LLVM libc." OFF)
```

Comment:
I would like to avoid introducing a CMake setting like this. Instead, we should do something similar to the way we handle ABI choices and provide a pseudo-target that represents the libc we're building against. We can then have a multi-valued option like `LIBCXX_LIBC="llvm-libc|system-libc|musl|etc..."`.

Reply:
That was my original thought, but I don't think the different C libraries have enough granularity for this to be anything but a binary flag in practice. What this flag truly means is whether or not to use the LLVM `libc` built in-tree or not. It's basically the exact same use-case as the already present `LIBCXXABI_USE_LLVM_UNWINDER`. If someone installed the LLVM libc globally, then the existing logic would just pick it up as the `libc` without this flag.

Reply:
The benefit of that approach is that we can define how to build against arbitrary new C stdlibs very easily, and we avoid adding more logic than there already is in the `CMakeLists.txt`. I don't mind if we start with only two configurations for this flag, what matters to me is that we use the right approach to introduce this new functionality.

Reply:
We have a solution to the same problem for selecting the ABI library: `set(LIBCXX_SUPPORTED_ABI_LIBRARIES none libcxxabi system-libcxxabi libcxxrt libstdc++ libsupc++ vcruntime)`.

There's `libcxxabi` (in-tree) and `system-libcxxabi` (whatever libc++abi is installed on the system).

Reply:
Okay, then we should have `system system-llvm-libc llvm-libc`, where `system` is the current default (Just use whatever you find in the default include path).

Reply:
I'm not sure how `system-llvm-libc` would differ from `system`, but yeah that sounds about right.

Reply:
Not necessary for now, was just thinking how if we started adding new ones like `musl` or `glibc` then we'd need to disambiguate. Right now we probably just want `system` as the default and `llvm-libc` which mirrors this patch's binary selection well enough.

Reply:
I introduced `LIBCXX_LIBC` which currently accepts `system` or `llvm-libc`. `system` is currently no-op but that can be expanded later. In the future we may also want to introduce additional options e.g. `musl` replacing the existing `LIBCXX_HAS_MUSL_LIBC` option but I'd prefer to do it in a separate change. 

I'm not sure if it's better to set the `-nostdlibinc` and `-nolibc` on the libc side (as interface flags on the targets) or on the libc++ side (as interface flags on the proxy targets), I can see an argument for doing either way. Right now I set those on the libc side which has the upside of not needing to duplicate those flags in libunwind and libc++abi which should eventually support LLVM libc as well.

In the future we'll probably want to move `HandleLibC.cmake` to the top-level `cmake/Modules` directory so they can be shared by libc++abi and libunwind, although we'll need to rethink the target and variables names because the `LIBCXX_` and `libcxx-` prefix is undesirable (maybe we can use `RUNTIMES_` and `runtimes-`)?